neurom/view/_dendrogram.py

```diff
@@ -27,280 +27,165 @@
 # SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 '''Dendrogram helper functions and class'''
+from collections import namedtuple
 
-from copy import deepcopy
-import sys
 import numpy as np
 
-from neurom.core import Tree, Neurite
+from neurom import NeuriteType
+from neurom.core import Neurite
 from neurom.core.dataformat import COLS
 
 
-def _n_terminations(tree):
-    '''Get the number of terminations in a tree'''
-    return sum(1 for _ in tree.ileaf())
+class Dendrogram:
+    '''Dendrogram'''
 
+    def __init__(self, neurom_section, dendrogram_root=None):
+        '''Dendrogram for NeuroM section tree.
 
-def _max_recursion_depth(obj):
-    ''' Estimate recursion depth, which is defined as the number of nodes in a tree
-    '''
-    neurites = obj.neurites if hasattr(obj, 'neurites') else [obj]
-
-    return max(sum(1 for _ in neu.iter_sections()) for neu in neurites)
-
-
-def _total_rectangles(tree):
-    '''
-    Calculate the total number of segments that are required
-    for the dendrogram. There is a vertical line for each segment
-    and two horizontal line at each branching point
-    '''
-    return sum(len(sec.children) + sec.points.shape[0] - 1
-               for sec in tree.iter_sections())
-
-
-def _n_rectangles(obj):
-    '''
-    Calculate the total number of rectangles with respect to
-    the type of the object
-    '''
-    return sum(_total_rectangles(neu) for neu in obj.neurites) \
-        if hasattr(obj, 'neurites') else _total_rectangles(obj)
-
-
-def _square_segment(radius, origin):
-    '''Vertices for a square
-    '''
-    return np.array(((origin[0] - radius, origin[1] - radius),
-                     (origin[0] - radius, origin[1] + radius),
-                     (origin[0] + radius, origin[1] + radius),
-                     (origin[0] + radius, origin[1] - radius)))
-
-
-def _vertical_segment(old_offs, new_offs, spacing, radii):
-    '''Vertices for a vertical rectangle
-    '''
-    return np.array(((new_offs[0] - radii[0], old_offs[1] + spacing[1]),
-                     (new_offs[0] - radii[1], new_offs[1]),
-                     (new_offs[0] + radii[1], new_offs[1]),
-                     (new_offs[0] + radii[0], old_offs[1] + spacing[1])))
-
-
-def _horizontal_segment(old_offs, new_offs, spacing, diameter):
-    '''Vertices of a horizontal rectangle
+        Args:
+            neurom_section: NeuroM section tree.
+            dendrogram_root: root of dendrogram. This is a service arg, please don't set it on
+            your own. It is used to track cycles in ``neurom_section``.
+        '''
+        if dendrogram_root is None:
+            dendrogram_root = self
+            dendrogram_root.processed_section_ids = []
+        if neurom_section.id in dendrogram_root.processed_section_ids:
```

Comment:
please use an assert for these kind of checks.

Also, is this necessary? I'm not sure how a cycle would happen - without the check, the `dendrogram_root` param wouldn't be necessary, and I think the logic would look neater?

Reply:
In the previous version we had `sys.setrecursionlimit(max_depth)`. I thought to replace this recursion limit with the cycles check. If the recursion limit didn't target cycles validation then I will remove my cycles check.

Reply:
This is the only major issue left to discuss. I would be happy to remove the cycles check but it would be nice to have a guarantee that we can't receive a cycled Neurite tree. Otherwise we need this check. I don't know where else to put it because we need this check per dendrogram tree instance. So each node needs an access to the tree instance. `dendrogram_root` is this instance.

Reply:
@wizmer Do we check that the loaded morphology does is not cycled? By the loaded morphology I mean `neurom.load_neuron('morph.swc')`. I don't see that we do it but I might overlooked it.

Reply:
For the structure of the file disallows cycles.  For SWC, we don't support non-connected components, so I think that's where it wouldn't load properly.  I can't remember for h5.

Reply:
I don't think the `sys.setrecursionlimit(max_depth)` was for validation, but probably because without it it would crash for neurites with too many descendents. 

I just tried on SWC with the following file and NeuroM did not complain although it should have, scary:
```
    1 1  0  0 0 1. -1
    2 3  0  0 0 1.  1
    3 3  0  5 0 1.  2
    4 3 -5  5 0 0.  3
    5 3  6  5 0 0.  3
    6 2  0  0 0 1.  6  # <-- cyclic point
    7 2  0 -4 0 1.  6
    8 2  6 -4 0 0.  7
    9 2 -5 -4 0 0.  7
```
The `mut_morphio` branch already raises on this file so I don't know if it's worth an implementation.